# Data Analysis

## Importing the libraries

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

## Dataset
### Showing the Data
Show the top columns of the dataset, which includes the data of each cancer tumour.

In [9]:
df = pd.read_csv('./data/snv-parse-real1.txt', sep='\t')
df

C:\Users\jAlej\AppData\Local\Temp\ipykernel_23388\184879827.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('./data/snv-parse-real1.txt', sep='\t')


Chr  START_POS_REF  END_POS_REF REF ALT   REF_MFVdVs   ALT_MFVdVs  \
0       1          13110        13110   G   A    G/NA/G/G/    A/NA/A/A/   
1       1          15015        15015   G   C   G/NA/NA/G/   C/NA/NA/C/   
2       1          16949        16949   A   C  NA/NA/NA/A/  NA/NA/NA/C/   
3       1          40552        40552   T   C  NA/NA/NA/T/  NA/NA/NA/C/   
4       1          46907        46907   T   C  NA/NA/NA/T/  NA/NA/NA/C/   
...    ..            ...          ...  ..  ..          ...          ...   
49315   Y       59028157     59028157   T   C  T/NA/NA/NA/  C/NA/NA/NA/   
49316   Y       59028833     59028833   G   T  G/NA/NA/NA/  T/NA/NA/NA/   
49317   Y       59030255     59030255   G   A  G/NA/NA/NA/  A/NA/NA/NA/   
49318   Y       59031439     59031439   T   A    T/NA/T/T/    A/NA/A/A/   
49319   Y       59031799     59031799   T   C  T/NA/NA/NA/  C/NA/NA/NA/   

      Sample_Name  FILTER_Mutect2  FILTER_Freebayes  FILTER_Vardict  \
0      icgc_cll-T            True             False           False   
1      icgc_cll-T            True             False           False   
2      icgc_cll-T           False             False           False   
3      icgc_cll-T           False             False           False   
4      icgc_cll-T           False             False           False   
...           ...             ...               ...             ...   
49315  icgc_cll-T            True             False           False   
49316  icgc_cll-T            True             False           False   
49317  icgc_cll-T            True             False           False   
49318  icgc_cll-T            True             False           False   
49319  icgc_cll-T            True             False           False   

       FILTER_Varscan  m2_MQ  f_MQMR  vs_SSC    vs_SPV   vd_SSF  vd_MSI  
0               False  41.91     NaN     2.0  0.522430  0.23427     2.0  
1               False  43.42     NaN     5.0  0.302390      NaN     NaN  
2                True    NaN     NaN    16.0  0.023282      NaN     NaN  
3                True    NaN     NaN    26.0  0.002231      NaN     NaN  
4                True    NaN     NaN    17.0  0.017670      NaN     NaN  
...               ...    ...     ...     ...       ...      ...     ...  
49315           False  54.17     NaN     NaN       NaN      NaN     NaN  
49316           False  56.34     NaN     NaN       NaN      NaN     NaN  
49317           False  56.37     NaN     NaN       NaN      NaN     NaN  
49318           False  54.38     NaN     2.0  0.525710  0.54983     1.0  
49319           False  58.63     NaN     NaN       NaN      NaN     NaN  

[49320 rows x 18 columns]

Use the method 'info' to show how many Non-null values are they, and what type of data object are they

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49320 entries, 0 to 49319
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Chr               49320 non-null  object 
 1   START_POS_REF     49320 non-null  int64  
 2   END_POS_REF       49320 non-null  int64  
 3   REF               49320 non-null  object 
 4   ALT               49320 non-null  object 
 5   REF_MFVdVs        49320 non-null  object 
 6   ALT_MFVdVs        49320 non-null  object 
 7   Sample_Name       49320 non-null  object 
 8   FILTER_Mutect2    49320 non-null  bool   
 9   FILTER_Freebayes  49320 non-null  bool   
 10  FILTER_Vardict    49320 non-null  bool   
 11  FILTER_Varscan    49320 non-null  bool   
 12  m2_MQ             8719 non-null   float64
 13  f_MQMR            7537 non-null   float64
 14  vs_SSC            39668 non-null  float64
 15  vs_SPV            39668 non-null  float64
 16  vd_SSF            16076 non-null  float6

Use the method 'describe' to see the distribution of the numerical values and statistics data. Good to look at min/max values to see any outliers that must put attention on, etc.

In [5]:
df.describe()

START_POS_REF   END_POS_REF        m2_MQ       f_MQMR        vs_SSC  \
count   4.932000e+04  4.932000e+04  8719.000000  7537.000000  39668.000000   
mean    6.509624e+07  6.509624e+07    54.510235    49.345667     15.221009   
std     5.462033e+07  5.462033e+07     7.070806    15.615680     17.291468   
min     3.000000e+00  3.000000e+00    10.130000     0.000000      0.000000   
25%     2.151707e+07  2.151707e+07    50.680000    43.193500      8.000000   
50%     5.135118e+07  5.135118e+07    57.480000    57.583300     12.000000   
75%     9.623363e+07  9.623363e+07    60.000000    60.000000     17.000000   
max     2.492405e+08  2.492405e+08    62.650000    70.000000    255.000000   

             vs_SPV        vd_SSF        vd_MSI  
count  39668.000000  16076.000000  16076.000000  
mean       0.118016      0.101967      4.169232  
std        0.169660      0.186070      5.324648  
min        0.000000      0.000000      0.000000  
25%        0.018322      0.000110      1.000000  
50%        0.051464      0.024045      2.000000  
75%        0.151600      0.117655      4.000000  
max        1.000000      1.000000     56.000000

The following method allows us to also see the number of null values of each column.

We see that m2_MQ and f_MQMR have simimlar missing values (which makes sense beacuse both tell about the mapping quality of the genome), then vs_SSC/SPV have also similar Nan values because both are related to varScan, and finally vd_SSF/MSI have similar NAN values because both are related to varDict. 

Possible reasons: 
- If there is high error in a read or the coverage was too low, then maybe outputs were not released.
- It could be related to how algorithms work and that in specific cases they decide to not output a given value.
- Maybe the researcher deleted those rows when preprocessing?
- Maybe if an algorithm didn't stored any SNV position, then skip the data. Maybe they provide metrics at variant side.
- Maybe all of the above.

In [6]:
df.isnull().sum()

Chr                     0
START_POS_REF           0
END_POS_REF             0
REF                     0
ALT                     0
REF_MFVdVs              0
ALT_MFVdVs              0
Sample_Name             0
FILTER_Mutect2          0
FILTER_Freebayes        0
FILTER_Vardict          0
FILTER_Varscan          0
m2_MQ               40601
f_MQMR              41783
vs_SSC               9652
vs_SPV               9652
vd_SSF              33244
vd_MSI              33244
dtype: int64

### Analysis

Print the columns that compose the dataset

In [7]:
columns = df.columns.tolist()
columns

['Chr',
 'START_POS_REF',
 'END_POS_REF',
 'REF',
 'ALT',
 'REF_MFVdVs',
 'ALT_MFVdVs',
 'Sample_Name',
 'FILTER_Mutect2',
 'FILTER_Freebayes',
 'FILTER_Vardict',
 'FILTER_Varscan',
 'm2_MQ',
 'f_MQMR',
 'vs_SSC',
 'vs_SPV',
 'vd_SSF',
 'vd_MSI']

To be able to see the rows that have null values, we can use the following method

In [9]:
df[df.isnull().any(axis=1)]

Chr  START_POS_REF  END_POS_REF REF ALT   REF_MFVdVs   ALT_MFVdVs  \
0       1          10303        10303   C   T  NA/NA/NA/C/  NA/NA/NA/T/   
1       1          16949        16949   A   C   A/NA/NA/A/   C/NA/NA/C/   
2       1          20628        20628   A   G    NA/A/A/A/    NA/G/G/G/   
3       1          30489        30489   G   A  NA/NA/NA/G/  NA/NA/NA/A/   
4       1          30808        30808   A   G  NA/NA/NA/A/  NA/NA/NA/G/   
...   ...            ...          ...  ..  ..          ...          ...   
3535    9      139138662    139138662   G   A  G/NA/NA/NA/  A/NA/NA/NA/   
3536    9      140378948    140378948   G   A  G/NA/NA/NA/  A/NA/NA/NA/   
3537    9      141046294    141046294   C   T  C/NA/NA/NA/  T/NA/NA/NA/   
3538    9      141062718    141062718   T   C  T/NA/NA/NA/  C/NA/NA/NA/   
3539    9      141081667    141081667   G   T  G/NA/NA/NA/  T/NA/NA/NA/   

          Sample_Name  FILTER_Mutect2  FILTER_Freebayes  FILTER_Vardict  \
0     icgc_cll_tumour           False             False           False   
1     icgc_cll_tumour           False             False           False   
2     icgc_cll_tumour           False             False            True   
3     icgc_cll_tumour           False             False           False   
4     icgc_cll_tumour           False             False           False   
...               ...             ...               ...             ...   
3535  icgc_cll_tumour            True             False           False   
3536  icgc_cll_tumour            True             False           False   
3537  icgc_cll_tumour            True             False           False   
3538  icgc_cll_tumour            True             False           False   
3539  icgc_cll_tumour            True             False           False   

      FILTER_Varscan  m2_MQ  f_MQMR  vs_SSC    vs_SPV   vd_SSF  vd_MSI  
0               True    NaN     NaN     5.0  0.277780      NaN     NaN  
1               True  22.92     NaN    20.0  0.008314      NaN     NaN  
2              False    NaN     NaN     8.0  0.139060  0.00469     2.0  
3               True    NaN     NaN    13.0  0.046735      NaN     NaN  
4               True    NaN     NaN    19.0  0.012309      NaN     NaN  
...              ...    ...     ...     ...       ...      ...     ...  
3535           False  60.00     NaN     NaN       NaN      NaN     NaN  
3536           False  60.00     NaN     NaN       NaN      NaN     NaN  
3537           False  58.97     NaN     NaN       NaN      NaN     NaN  
3538           False  58.84     NaN     NaN       NaN      NaN     NaN  
3539           False  46.62     NaN     NaN       NaN      NaN     NaN  

[3470 rows x 18 columns]

## Understanding the data

The following code will be used to understand what each columns mean and what do their value represent

In [8]:
columns = df.columns.tolist()
print(columns)
df[df.isnull().any(axis=1)]
df.head()

['Chr', 'START_POS_REF', 'END_POS_REF', 'REF', 'ALT', 'REF_MFVdVs', 'ALT_MFVdVs', 'Sample_Name', 'FILTER_Mutect2', 'FILTER_Freebayes', 'FILTER_Vardict', 'FILTER_Varscan', 'm2_MQ', 'f_MQMR', 'vs_SSC', 'vs_SPV', 'vd_SSF', 'vd_MSI']


Chr  START_POS_REF  END_POS_REF REF ALT   REF_MFVdVs   ALT_MFVdVs  \
0   1          13110        13110   G   A    G/NA/G/G/    A/NA/A/A/   
1   1          15015        15015   G   C   G/NA/NA/G/   C/NA/NA/C/   
2   1          16949        16949   A   C  NA/NA/NA/A/  NA/NA/NA/C/   
3   1          40552        40552   T   C  NA/NA/NA/T/  NA/NA/NA/C/   
4   1          46907        46907   T   C  NA/NA/NA/T/  NA/NA/NA/C/   

  Sample_Name  FILTER_Mutect2  FILTER_Freebayes  FILTER_Vardict  \
0  icgc_cll-T            True             False           False   
1  icgc_cll-T            True             False           False   
2  icgc_cll-T           False             False           False   
3  icgc_cll-T           False             False           False   
4  icgc_cll-T           False             False           False   

   FILTER_Varscan  m2_MQ  f_MQMR  vs_SSC    vs_SPV   vd_SSF  vd_MSI  
0           False  41.91     NaN     2.0  0.522430  0.23427     2.0  
1           False  43.42     NaN     5.0  0.302390      NaN     NaN  
2            True    NaN     NaN    16.0  0.023282      NaN     NaN  
3            True    NaN     NaN    26.0  0.002231      NaN     NaN  
4            True    NaN     NaN    17.0  0.017670      NaN     NaN

## Understanding the columns

Breakdown of each column

---

## **1. Genomic Position Information**
| Column | Description | Example Value |
|--------|------------|--------------|
| **Chr** | Chromosome number where the SNV is located. | `1` |
| **START_POS_REF** | Start position of the SNV in the reference genome. | `13110` |
| **END_POS_REF** | End position of the SNV in the reference genome (same as `START_POS_REF` for SNVs because an SNV is a single nucleotide that was modified). | `13110` |

- **Format:** Integer (`1, 2, ..., 22, X, Y` for chromosomes)
- **Purpose:** Identifies where in the genome this SNV occurs.

---

## **2. Variant Information**
| Column | Description | Example Value |
|--------|------------|--------------|
| **REF** | Reference nucleotide (what is expected in a normal genome). | `G` |
| **ALT** | Alternate nucleotide (mutation found in the tumor sample). | `A` |

- **Format:** Single-character DNA base (`A, T, C, G`)
- **Purpose:** Shows how the sequence has changed in cancer cells.

---

## **3. Allele Information Across Algorithms**
| Column | Description | Example Value |
|--------|------------|--------------|
| **REF_MFVdVs** | Reference allele calls from all four algorithms: MuTect2 (M), FreeBayes (F), VarDict (Vd), and VarScan (Vs). | `G/NA/G/G/` |
| **ALT_MFVdVs** | Alternate allele calls from all four algorithms: MuTect2 (M), FreeBayes (F), VarDict (Vd), and VarScan (Vs). | `A/NA/A/A/` |

In the example above, it is telling us that MuTect2, VarDict, and VarScan expected the reference nucleotie (what is expected in normal genome) to be 'G', but the alterate allele was 'A', meaning it is a SNV. On the other hand, FreeBayes didn't disagreed that a mutation exists.

- **Format:** Slash-separated (`M/F/Vd/Vs/`), `NA` means the algorithm didn't detect a variant at this position.
- **Purpose:** Shows how each SNV caller interpreted the mutation.

---

## **4. Sample Information**
| Column | Description | Example Value |
|--------|------------|--------------|
| **Sample_Name** | The tumor sample ID. | `icgc_cll-T` |

- **Format:** String (sample name)
- **Purpose:** Identifies which tumor sample this variant belongs to.

---

## **5. Variant Detection by Each Algorithm**
| Column | Description | Example Value |
|--------|------------|--------------|
| **FILTER_Mutect2** | `TRUE` if MuTect2 passed the variant as significant, `FALSE` otherwise. | `TRUE` |
| **FILTER_Freebayes** | `TRUE` if FreeBayes passed the variant, `FALSE` otherwise. | `FALSE` |
| **FILTER_Vardict** | `TRUE` if VarDict passed the variant, `FALSE` otherwise. | `FALSE` |
| **FILTER_Varscan** | `TRUE` if VarScan passed the variant, `FALSE` otherwise. | `FALSE` |

- **Format:** Boolean (`TRUE` or `FALSE`)
- **Purpose:** Indicates whether each algorithm considers the SNV a real mutation.

---

## **6. Variant-Specific Features (Algorithm-Specific Quality Scores)**
| Column | Description | Example Value |
|--------|------------|--------------|
| **m2_MQ** | Mapping quality from **MuTect2** (how confidently the read aligns to the genome). | `41.91` |
| **f_MQMR** | Mapping quality from **FreeBayes**. | `NA` |
| **vs_SSC** | Strand bias from **VarScan** (how evenly reads support the mutation in both DNA strands). | `2` |
| **vs_SPV** | Statistical p-value from **VarScan** (significance of the mutation). | `0.52243` |
| **vd_SSF** | Somatic status score from **VarDict**. | `0.23427` |
| **vd_MSI** | Microsatellite instability (MSI) from **VarDict**. | `2` |

- **Format:** Float or `NA` (if the algorithm didn't calculate it).
- **Purpose:** Algorithm-specific quality metrics for filtering out false positives.

---

## **Summary**
- **Columns 1-3** → Where the variant is in the genome.
- **Columns 4-5** → What the mutation is.
- **Columns 6-7** → How each algorithm sees the reference and alternate alleles.
- **Column 8** → Which tumor sample the variant belongs to.
- **Columns 9-12** → Whether each algorithm validated this SNV.
- **Columns 13-18** → Algorithm-specific scores for quality assessment.

## Results
### Showing and Understanding the data
Show the top columns of the dataset in the '.bed' represent the true somatic SNVs

In [19]:
df = pd.read_csv('./real1/real1_truth.bed', sep='\t', header=None)
df

0          1          2
0     1    2180985    2180985
1     1    5035185    5035185
2     1    8881322    8881322
3     1    8929624    8929624
4     1    9196716    9196716
...  ..        ...        ...
1314  X  142275497  142275497
1315  X  144737796  144737796
1316  Y    6407994    6407994
1317  Y   15154839   15154839
1318  Y   15681833   15681833

[1319 rows x 3 columns]

Each BED file contains **three columns**, and each row in a BED file represents a **confirmed somatic SNV**.

| Column  | Description                                | Example  |
|---------|--------------------------------------------|---------|
| **Chr** | Chromosome number                         | `1`     |
| **Start** | Start position of the SNV               | `13110` |
| **End** | End position of the SNV *(same as Start for SNVs)* | `13110` |